
diff --git a/src/login/logind-button.c b/src/login/logind-button.c
index 9ec235a..fcc0f07 100644
--- a/src/login/logind-button.c
+++ b/src/login/logind-button.c
@@ -13,6 +13,7 @@
 #include "missing_input.h"
 #include "string-util.h"
 #include "util.h"
+#include "async.h"
 
 #define CONST_MAX4(a, b, c, d) CONST_MAX(CONST_MAX(a, b), CONST_MAX(c, d))
 
@@ -63,7 +64,7 @@ void button_free(Button *b) {
                 /* If the device has been unplugged close() returns
                  * ENODEV, lets ignore this, hence we dont use
                  * safe_close() */
-                (void) close(b->fd);
+                (void) asynchronous_close(b->fd);
 
         free(b->name);
         free(b->seat);
@@ -330,7 +331,7 @@ int button_open(Button *b) {
         if (r < 0)
                 return log_warning_errno(r, "Failed to determine whether input device %s is relevant to us: %m", p);
         if (r == 0)
-                return log_debug_errno(SYNTHETIC_ERRNO(EADDRNOTAVAIL),
+                return b->fd = TAKE_FD(fd), log_debug_errno(SYNTHETIC_ERRNO(EADDRNOTAVAIL),
                                        "Device %s does not expose keys or switches relevant to us, ignoring.", p);
 
         if (ioctl(fd, EVIOCGNAME(sizeof name), name) < 0)

